Training/Training/Training.c

```diff
@@ -27,34 +26,37 @@
 
 /// <summary>Function which checks if a given string is palindrome or not</summary>
 int PalindromeChecker (char* phrase) {
-   int i = 0, j = 0, length;
-   char temp[MAX] = "";
+   int i = 0, j, length;
    length = (int)strlen (phrase);
-   if (length > MAX - 2) {
+   j = length - 1;
+   if (length > MAX - 2)
       return INVALID;
-   }
-   // Eliminating special characters by moving the remaining character into a temporary array
-   while (i < length) {
-      if (isalnum (phrase[i]) != 0)
-         temp[j++] = phrase[i];
-      i++;
-   }
-   temp[j] = '\0';
-   // Checking invalid text
-   if (j == 0)
-      return INVALID;
-   i = 0;
-   while (i < j / 2) {
-      if (toupper (temp[i]) != toupper (temp[j - 1 - i]))
-         return NOTPALINDROME;
-      i++;
+   // Checking if valid characters match with the valid characters  
+   while (i < length / 2 && j>length / 2) {
+      if (isalnum (phrase[i]) == 0)
+         i++;
+      else if (isalnum (phrase[j]) == 0)
+         j--;
+      else
+         if (isalnum (phrase[i]) != 0 && isalnum (phrase[j]) != 0) {
+            if (toupper (phrase[i]) != toupper (phrase[j]))
+               return NOTPALINDROME;
+            else {
+               i++;
+               j--;
+            }
+         }
    }
    return PALINDROME;

```

Comment:
Returns PALINDROME if empty string or string consisting of only special characters is given as input. But it should check whether the input contains any valid characters to compare and only return palindrome if the valid string is a palindrome.